kube-core/src/schema/transform_one_of.rs

```diff
@@ -0,0 +1,181 @@
+/// enum of the same type.
+///
+/// This will return early without modifications unless:
+/// - There are `oneOf` subschemas (not empty).
+/// - Each subschema contains an `enum` or `const`.
+///
+/// Subschemas must define a type, and they must be the same for all.
+///
+/// NOTE: This should work regardless of whether other hoisting has been
+/// performed or not.
+#[derive(Debug, Clone)]
+pub struct OneOfSchemaRewriter;
+
+impl Transform for OneOfSchemaRewriter {
+    fn transform(&mut self, transform_schema: &mut schemars::Schema) {
+        // Apply this (self) transform to all subschemas
+        schemars::transform::transform_subschemas(self, transform_schema);
+
+        let mut schema: SchemaObject = match serde_json::from_value(transform_schema.clone().to_value()).ok()
+        {
+            // TODO (@NickLarsenNZ): At this point, we are seeing duplicate `title` when printing schema as json.
+            // This is because `title` is specified under both `extensions` and `other`.
+            Some(schema) => schema,
+            None => return,
+        };
+
+        hoist_one_of_enum_with_unit_variants(&mut schema);
+
+        // Convert back to schemars::Schema
+        if let Ok(schema) = serde_json::to_value(schema) {
+            if let Ok(transformed) = serde_json::from_value(schema) {
+                *transform_schema = transformed;
+            }
+        }
+    }
+}
+
+fn hoist_one_of_enum_with_unit_variants(kube_schema: &mut SchemaObject) {
+    // Run some initial checks in case there is nothing to do
+    let SchemaObject {
+        subschemas: Some(subschemas),
+        ..
+    } = kube_schema
+    else {
+        return;
+    };
+
+    let SubschemaValidation {
+        any_of: None,
+        one_of: Some(one_of),
+    } = subschemas.deref_mut()
+    else {
+        return;
+    };
+
+    if one_of.is_empty() {
+        return;
+    }
+
+    // At this point, we can be reasonably sure we need to hoist the oneOf
+    // subschema enums and types up to the schema level, and unset the oneOf field.
+    // From here, anything that looks wrong will panic instead of return.
+    // TODO (@NickLarsenNZ): Return errors instead of panicking, leave panicking up to the
+    // infallible schemars::Transform
+
+    // Prepare to ensure each variant schema has a type
+    let mut types = one_of.iter().map(|schema| match schema {
+        Schema::Object(SchemaObject {
+            instance_type: Some(r#type),
+            ..
+        }) => r#type,
+        Schema::Object(untyped) => panic!("oneOf variants need to define a type: {untyped:#?}"),
+        Schema::Bool(_) => panic!("oneOf variants can not be of type boolean"),
+    });
+
+    // Get the first type
+    let variant_type = types.next().expect("at this point, there must be a type");
+    // Ensure all variant types match it
+    if types.any(|r#type| r#type != variant_type) {
+        panic!("oneOf variants must all have the same type");
+    }
```

Comment:
probably a sane unwrap for now. might need another line about why this is true in the code.

Reply:
Added extra explanation in 77af8e5d4dec04daa6ad904990edbd7b8c7b5f4b